### PR TITLE
Configurable audio codecs in VideoRoom plugin.

### PR DESCRIPTION
**Pull request by pallab-gain**
This pull request was inspired by recent PR #478 , and from our needs to support other audio codecs such as PCMU, iSAC 16K, and iSAC 32K. By default OPUS audio codec will be selected, but we can choose PCMU/8000 `( audiocodec=pcmu )` , iSAC/16000 `( audiocodec=isac16 )`, and iSAC/32000 `( audiocodec=isac32 )` when creating a room request. This extended codec support inherits same limitation as PR #478; i.e. it will not work in post-processing, and recording right now.

### CODE REVIEW

**Comment by lminiero** [comment]
Thanks for working on this! I'll add some comments inline.

File: conf/janus.plugin.videoroom.cfg.sample [hunk 1]
```diff
@@ -18,5 +18,6 @@ publishers = 6
 bitrate = 128000
 fir_freq = 10
 ;videocodec = vp8
+;audiocodec = opus
```
Comment (lminiero) [inline conf/janus.plugin.videoroom.cfg.sample]: Can you put the audiocodec stuff before the videocodec stuff? Throughout all the code I mean (config as in here, when parsing, when printing, etc.). This is the convention we follow in all the gateway.

File: plugins/janus_videoroom.c [hunk 2]
```diff
@@ -12,7 +12,7 @@
  * different scenarios, ranging from a simple webinar (one speaker, several
  * listeners) to a fully meshed video conference (each peer sending and
  * receiving to and from all the others).
- * 
+ *
```
Comment (lminiero) [inline plugins/janus_videoroom.c]: Not sure why you removed all those extra spaces, but ok :smile: 
Did the IDE autofix those for you?

File: plugins/janus_videoroom.c [hunk 3]
```diff
@@ -251,6 +251,15 @@ typedef enum janus_videoroom_videocodec {
 	JANUS_VIDEOROOM_VP9,	/* Publishers will have to use VP9 */
 	JANUS_VIDEOROOM_H264	/* Publishers will have to use H264 */
 } janus_videoroom_videocodec;
+
+typedef enum janus_videoroom_audiocodec {
+	JANUS_VIDEOROOM_OPUS,	/* Publishers will have to use OPUS */
+	JANUS_VIDEOROOM_ISAC_32K,	/* Publishers will have to use ISAC 32K */
+	JANUS_VIDEOROOM_ISAC_16K,	/* Publishers will have to use ISAC 16K */
+	JANUS_VIDEOROOM_PCMU	/* Publishers will have to use PCMU 8K */
```
Comment (lminiero) [inline plugins/janus_videoroom.c]: Can you add PCMA (a-Law) as well? It's another codec all browsers should support.

File: plugins/janus_videoroom.c [hunk 4]
```diff
@@ -383,11 +413,30 @@ typedef struct janus_videoroom_rtp_relay_packet {
...
 		"c=IN IP4 1.1.1.1\r\n" \
 		"a=%s\r\n"							/* Media direction */ \
-		"a=rtpmap:%d opus/48000/2\r\n"		/* Opus payload type */
+		"a=rtpmap:%d ISAC/16000\r\n"		/* ISAC16_PT payload type */
+
+#define sdp_a_template_pcmu \
+		"m=audio 1 RTP/SAVPF %d\r\n"		/* Opus payload type */ \
+		"c=IN IP4 1.1.1.1\r\n" \
+		"a=%s\r\n"							/* Media direction */ \
+		"a=rtpmap:%d PCMU/8000\r\n"		/* ISAC16_PT payload type */
```
Comment (lminiero) [inline plugins/janus_videoroom.c]: PCMU_PT

File: plugins/janus_videoroom.c [hunk 5]
```diff
@@ -699,10 +765,11 @@ int janus_videoroom_init(janus_callbacks *callback, const char *config_path) {
 			janus_mutex_lock(&rooms_mutex);
 			g_hash_table_insert(rooms, GUINT_TO_POINTER(videoroom->room_id), videoroom);
 			janus_mutex_unlock(&rooms_mutex);
-			JANUS_LOG(LOG_VERB, "Created videoroom: %"SCNu64" (%s, %s, %s codec, secret: %s, pin: %s)\n",
+			JANUS_LOG(LOG_VERB, "Created videoroom: %"SCNu64" (%s, %s, %s %s codec, secret: %s, pin: %s)\n",
```
Comment (lminiero) [inline plugins/janus_videoroom.c]: Maybe a prettier way of showing both codecs here: ```%s/%s codecs```?

File: plugins/janus_videoroom.c [hunk 6]
```diff
@@ -1244,10 +1347,11 @@ struct janus_plugin_result *janus_videoroom_handle_message(janus_plugin_session
 		videoroom->destroyed = 0;
 		janus_mutex_init(&videoroom->participants_mutex);
 		videoroom->participants = g_hash_table_new(NULL, NULL);
-		JANUS_LOG(LOG_VERB, "Created videoroom: %"SCNu64" (%s, %s, %s codec, secret: %s, pin: %s)\n",
+		JANUS_LOG(LOG_VERB, "Created videoroom: %"SCNu64" (%s, %s, %s %s codec, secret: %s, pin: %s)\n",
```
Comment (lminiero) [inline plugins/janus_videoroom.c]: Same as above.

File: plugins/janus_videoroom.c [hunk 7]
```diff
@@ -3535,11 +3662,26 @@ static void *janus_videoroom_handler(void *data) {
 				}
 				sdp_parser_free(parser);
 				JANUS_LOG(LOG_VERB, "The publisher %s going to send an audio stream\n", audio ? "is" : "is NOT");
-				int opus_pt = 0, vp8_pt = 0, vp9_pt = 0, h264_pt = 0;
+				int opus_pt = 0, isac32_pt = 0, isac16_pt = 0, pcmu_pt = 0, vp8_pt = 0, vp9_pt = 0, h264_pt = 0;
 				if(audio) {
 					JANUS_LOG(LOG_VERB, "  -- Will answer with media direction '%s'\n", audio_mode);
 					opus_pt = janus_get_opus_pt(msg->sdp);
-					JANUS_LOG(LOG_VERB, "  -- -- Opus payload type is %d\n", opus_pt);
+					if(opus_pt > 0){
```
Comment (lminiero) [inline plugins/janus_videoroom.c]: Code convention: space between those bracket,  ```if(opus_pt > 0) {```

File: plugins/janus_videoroom.c [hunk 8]
```diff
@@ -3564,10 +3706,55 @@ static void *janus_videoroom_handler(void *data) {
 					b = (int)(videoroom->bitrate/1000);
 				char sdp[1280], audio_mline[256], video_mline[512], data_mline[256];
 				if(audio) {
-					g_snprintf(audio_mline, 256, sdp_a_template,
-						opus_pt,						/* Opus payload type */
-						audio_mode,						/* The publisher gets a recvonly or inactive back */
-						opus_pt); 						/* Opus payload type */
+					switch (videoroom->acodec) {
```
Comment (lminiero) [inline plugins/janus_videoroom.c]: Code convention: no space between method and arguments, ```switch(videoroom->acodec)``` (same for ```if``` etc.)

File: plugins/janus_videoroom.c [hunk 9]
```diff
@@ -3700,11 +3887,58 @@ static void *janus_videoroom_handler(void *data) {
 
 				/* Now turn the SDP into what we'll send subscribers, using the static payload types for making switching easier */
 				if(audio) {
-					g_snprintf(audio_mline, 256, sdp_a_template,
-						OPUS_PT,						/* Opus payload type */
-						/* Subscribers gets a sendonly or inactive back */
-						strcmp(audio_mode, "inactive") ? "sendonly" : "inactive",
-						OPUS_PT); 						/* Opus payload type */
+						switch ( videoroom->acodec ) {
```
Comment (lminiero) [inline plugins/janus_videoroom.c]: Code convention: no spaces between brackets and arguments: ```switch(videoroom->acodec) {```

File: plugins/janus_videoroom.c [hunk 10]
```diff
@@ -3700,11 +3887,58 @@ static void *janus_videoroom_handler(void *data) {
...
+								case JANUS_VIDEOROOM_OPUS:
+									if(opus_pt < 0){
+										audio_mline[0] = '\0';
+									}else{
```
Comment (lminiero) [inline plugins/janus_videoroom.c]: Code convention: spaces there, ```} else {```

File: utils.c [hunk 11]
```diff
@@ -232,6 +232,102 @@ int janus_get_opus_pt(const char *sdp) {
...
+					return pt;
+				}
+			}
+			*next = '\n';
+		}
+		line = next ? (next+1) : NULL;
+	}
+	return -3;
+}
+
```
Comment (lminiero) [inline utils.c]: I'll probably have to come up with a single method that can get the payload type according to the provided codec name, there are too many of them now...

**Comment by lminiero** [comment]
As a side note, if you haven't signed our [CLA](https://cla.conf.meetecho.com/janus/) yet, please do so before we can merge your contribution.

File: plugins/janus_videoroom.c (hunk 3 above)
Comment (pallab-gain) [inline plugins/janus_videoroom.c]: Thanks. I will add PCMA ( a-Law ), and will test, and send you another request with fixes, and suggestions that you mentioned in your inline comments. Thanks for your time.

Comment (lminiero) [inline plugins/janus_videoroom.c]: Thanks!

File: plugins/janus_videoroom.c (hunk 8 above)
Comment (pallab-gain) [inline plugins/janus_videoroom.c]: I am so sorry. I don't use IDE, my custom editor ruined it. I will be more careful from now on.

Comment (lminiero) [inline plugins/janus_videoroom.c]: You didn't ruin anything, these are actually fixes! :wink:

**Comment by lminiero** [comment]
Thanks for the fixes: there still are a few indentations and code conventions to fix, but I can do that myself after merging so no need to worry about that. Apart from that, the patch looks fine to me!

As soon as you can confirm you signed the CLA I'll merge it in master, thanks for working on this!

**Comment by pallab-gain** [comment]
Sorry, I am still working on indentations, and code conventions fixing to make sure we have to do minimum works on this part once we accept the merge request. I am still giving PCMA few more testing, and working on indentation. My master is still few commits behind. I will comment as soon as it is ready. Sorry, if this confuse you. This is the first time I am trying to work with open source project. Bu, I am learning very fast.  :smile:

**Comment by lminiero** [comment]
No problem at all, your contribution has been very welcome and you're doing a great job!

**Comment by pallab-gain** [comment]
I tested all the audio codecs, and fixes few indentations, and tried to follow code conventions as you suggest. When you get time please check, and let me know if we need to push some few more fixes ? 

I also submited the CLA form.

**Comment by lminiero** [comment]
Looks fine, thanks for the updates! Merging right away then.

**Pull request merged**